### ISSUE TITLE
E-cash note was replaced in DB, this should never happen!

### ISSUE THREAD

**Issue by douglaz** [fedimint/fedimint#2543]
https://github.com/douglaz/fedimint/pull/new/note_replace_error

```log
$ fedimint-load-test-tool --users 10 run-economy  --generate-invoice-with cln 
2023-05-16T18:06:11.709661Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:11.760534Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:11.813822Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:11.866876Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:11.920666Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:11.973952Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.027095Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.080339Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.133590Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.186591Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.239633Z  INFO fedimint_client::sm::executor: Starting state machine executor task
2023-05-16T18:06:12.241746Z  INFO fedimint_load_test_tool: Reissuing initial notes: initial 1000000 msat, per user: 1100 msat
2023-05-16T18:06:20.762722Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(7638980573973adb7459619d44a066b4309ecb12df5fcfd5ab98d053c8464338520d2970a9d81bb3a1153f51f33498adff9aab9500d83aae9bbd4a62ee09802a)), Signature(G1Affine { x: 0x18e8d178a82c8a45ec3583136bc67cb61ab904bbfed13d7685e8dfe03f69914ee2f84c7f70dc5a1635d361d4e95e5070, y: 0x0f31c41acab7c951fb458bd7e10c693ae34c1ff7c48bb865f3f1dfa88adf0272e491cb461ef85e610bc0c08d235e8b6a, infinity: Choice(0) })), spend_key: KeyPair(#0194841ecb6d1f3a) }
2023-05-16T18:06:20.807750Z  INFO fedimint_load_test_tool: 1 msat: 2
2023-05-16T18:06:20.807774Z  INFO fedimint_load_test_tool: 2 msat: 3
2023-05-16T18:06:20.807781Z  INFO fedimint_load_test_tool: 4 msat: 2
2023-05-16T18:06:20.807788Z  INFO fedimint_load_test_tool: 8 msat: 2
2023-05-16T18:06:20.807794Z  INFO fedimint_load_test_tool: 16 msat: 2
2023-05-16T18:06:20.807800Z  INFO fedimint_load_test_tool: 32 msat: 2
2023-05-16T18:06:20.807807Z  INFO fedimint_load_test_tool: 64 msat: 3
2023-05-16T18:06:20.807813Z  INFO fedimint_load_test_tool: 128 msat: 2
2023-05-16T18:06:20.807819Z  INFO fedimint_load_test_tool: 256 msat: 2
2023-05-16T18:06:20.807825Z  INFO fedimint_load_test_tool: 512 msat: 3
2023-05-16T18:06:20.807831Z  INFO fedimint_load_test_tool: 1024 msat: 31
2023-05-16T18:06:20.807837Z  INFO fedimint_load_test_tool: 2048 msat: 1
2023-05-16T18:06:20.807843Z  INFO fedimint_load_test_tool: 4096 msat: 3
2023-05-16T18:06:20.807848Z  INFO fedimint_load_test_tool: 8192 msat: 2
2023-05-16T18:06:20.807854Z  INFO fedimint_load_test_tool: 16384 msat: 3
2023-05-16T18:06:20.807860Z  INFO fedimint_load_test_tool: 32768 msat: 1
2023-05-16T18:06:20.807866Z  INFO fedimint_load_test_tool: 65536 msat: 3
2023-05-16T18:06:20.807872Z  INFO fedimint_load_test_tool: 131072 msat: 3
2023-05-16T18:06:20.807878Z  INFO fedimint_load_test_tool: 262144 msat: 1

**Comment by douglaz** [comment]
> Have we seen this recently @douglaz?

No, this isn't common, it's a corner case that I was only able to reproduce on that branch

**Comment by justinmoon** [comment]
@douglaz Should we close this?

**Comment by douglaz** [comment]
Got this again on some load test on v0.1:
```
2023-11-08T00:36:59.105094Z  INFO fedimint_client::sm::executor: Starting state machine executor task                                                                                                                         [73/1900]
2023-11-08T00:36:59.110145Z  INFO fedimint_load_test_tool: Current balance of 20480 msat already covers the minimum required of 10240 msat                                                                                             
2023-11-08T00:36:59.110174Z  INFO fedimint_load_test_tool: Reminting notes of denomination 1024 msat for 10 users, 1 notes per user (this may take a while if the number of users/notes is high)                                       
2023-11-08T00:36:59.179032Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(76c799f708f03eecee3000823b6670688404d6a96f0f9817b9724f51
6dece7c198d866ced9b8bc0dac3a8afbf48196fa99ddca4fea5e101aec70b07ad6e9448e)), Signature(G1Affine { x: 0x022da1c4acbbfd1aaf7e42d047551e074fc6bc883e4f8848b4d8bf600a2c3e3f75219cbb8a880eb1feffa061cd8e153d, y: 0x17a5d4cf1ad8c09dd357087cfa
fe773f44b6d3bd0245c379379a645c524944ff6ab0b9d005a0e8251028cb37e512b4fe, infinity: Choice(0) })), spend_key: KeyPair(#e34fc6a0cc58a36e) }                                                                                               
2023-11-08T00:36:59.183564Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(0b6e0f41f0b742a06a213ea59a2c94cf36023783890cba9492c5b43a
bc595f38f025baf10565d75f0fdf7acaba587adc6711a673a4183da1d558a5cc9075325d)), Signature(G1Affine { x: 0x0db0db92916d2f02f0820bebe08e3f6c18ef60714bb69fc8ed25d205c172f361ad88a1a7d86a14f1aa7054eb6da0ca3a, y: 0x19ba69946677859166c20f2414
4b0402cda3e5c6e021e7192f7196d6a6bd27fea6e8c27b5137932864fa1dc1306108df, infinity: Choice(0) })), spend_key: KeyPair(#0e38777e5e159a1f) }                                                                                               
2023-11-08T00:36:59.184294Z  WARN db: Database commit failed in an autocommit block curr_attempts=1                                                                                                                                    
2023-11-08T00:36:59.187810Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(5bd8187d0bf04415200b53a2826d5422637a889747db23e3f4a32a16
a1f8a7e3ac6acbba4743530146d6036a8f751d1b1217f17698cfcb957eed04a0d88e9a20)), Signature(G1Affine { x: 0x023ec39b5bc7f9f604e555efc40ccc9e82dc2506f3c1eb6a4047d3c576c73df0e9019dc663bc2e2f21ed583bff7f9c47, y: 0x09062316b2387ea64875e1bd76
9a7447ae191486ef14dbc0b4d8c40c0145abc1541d60be29033347688df4a5977cbf96, infinity: Choice(0) })), spend_key: KeyPair(#a3f5ada8e5c29074) }                                                                                               
2023-11-08T00:36:59.192033Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(11b3d43c8039b016d500e9f4e93edbc6bfab9cb4d0109f3e4f016710
2501c5abf0ba8dff07dc60a2e40d550864f25bdb1ea1a3d2d7efcc30b9e69415a8a55b78)), Signature(G1Affine { x: 0x1508c90801cca9ac4fb97c737945f679339c83eb8299b740d8e0a0af69cd601fe5b093d1ad65a0c880b5e9e5873e9822, y: 0x0a6e1d174588745a635608f240
b956a9fbd38816ee07d17396202571cd6b09aaec5b4829e3a09151648faec99562e91a, infinity: Choice(0) })), spend_key: KeyPair(#1b54056f92d8574c) }                                                                                               
2023-11-08T00:36:59.196288Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(b945165454b7d70b71a86bbab8393a0d1991cdbbccbe9603d11d38bd
b8c50d1ad88e4bf1c0639634bcf36dbea8b5121c29ed5bdb020e8384f8a139edd6979967)), Signature(G1Affine { x: 0x0a002dd9a039c568ed30c293d22ca66ff2d81487a739a2676753a576eb41404c32d04ff9e1988951436021d1de84f81f, y: 0x10d89bbcae41a49cfa216341ce
af1cc21ae2ec6b15263485b1d3a002387912ba5390780fa54f9cecd16712739b8ea4ba, infinity: Choice(0) })), spend_key: KeyPair(#54e3c2919801be0b) }                                                                                               
2023-11-08T00:36:59.200346Z  WARN db: Database commit failed in an autocommit block curr_attempts=2                                                                                                                                    
2023-11-08T00:36:59.201144Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(deac4f1343d81397c2336c390bbd636a28ba1e521557a3cd62da76f1
f58680a3e0d34b90f4da6d738233653f62bebd4e913e42745064582b3627b8d3deafdadd)), Signature(G1Affine { x: 0x094ce0e53ec28fe908c0e09ae19c6287b9b5d423ab46ab6b35fcafeb7076dd6da43bca3bb552c49947526b6467703c98, y: 0x00166b2c23dea68c8de4e29c6a
ab1f777dc451ce4f44848b7f3919a0fc050f1aa6ba1166be2783538063600ebc0652cf, infinity: Choice(0) })), spend_key: KeyPair(#c2c0677829f7b658) }                                                                                               
2023-11-08T00:36:59.205640Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(0ea5f52e53c68c709d34ab68b5a3418f3b1dad7319854086cc6721ce
a492f42588a39713066df4d5f61370adda973ed37a6bd9f367a7c1defe6b5051d602e646)), Signature(G1Affine { x: 0x02ed0109eaa2fc6eb61cc0938084ed31cc306a01116350e6696371ff709f550226527c57a54c1170b82a7d22dbfdcec9, y: 0x002f286fa66c531d642114a2a1
70edb0040c1fda2cd892d724111261d0722bd4fb1fbb5d11f1734c27fd8a1addb03d26, infinity: Choice(0) })), spend_key: KeyPair(#cf738889477737ff) }                                                                                               
2023-11-08T00:36:59.210007Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(e02c4a6ada2476102850083814634df86f93eff4df609e470f3676f3
74512b90a8c535279fc727595a81b33f99dcb1430f2b7949c0261332abfd3ce4c11bef2b)), Signature(G1Affine { x: 0x08fd4e26f680848cd25d80c6ce6775965d940bac649c584887de5649d85294a4d858f15ed3ebd13f1fcbdf0d6aa7be93, y: 0x094bfa893ca36b73c20737ca26
001cee3b17fac7b119f3c8c2430e9dce15d3122f994f9ad5252379ab23b02f6ae9b8a1, infinity: Choice(0) })), spend_key: KeyPair(#ae457a5e2048d982) }                                                                                               
2023-11-08T00:36:59.232162Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(0b6e0f41f0b742a06a213ea59a2c94cf36023783890cba9492c5b43a
bc595f38f025baf10565d75f0fdf7acaba587adc6711a673a4183da1d558a5cc9075325d)), Signature(G1Affine { x: 0x0db0db92916d2f02f0820bebe08e3f6c18ef60714bb69fc8ed25d205c172f361ad88a1a7d86a14f1aa7054eb6da0ca3a, y: 0x19ba69946677859166c20f2414
4b0402cda3e5c6e021e7192f7196d6a6bd27fea6e8c27b5137932864fa1dc1306108df, infinity: Choice(0) })), spend_key: KeyPair(#0e38777e5e159a1f) }                                                                                               
2023-11-08T00:37:01.502013Z ERROR fedimint_mint_client::output: E-cash note was replaced in DB, this should never happen! note=SpendableNote { note: Note(Nonce(XOnlyPublicKey(32f6b4527ee60e34951563c16c12e46b46004d3296782300bcfa2946
db6efbded4895b5cffaf076b8fd928633e63b572d2743193fd7d891c6731377117159a3a)), Signature(G1Affine { x: 0x1986157de87d53f56be680c7913580284795105689d28985af7deaae2a84c4b53fc6afd4558450c7adbaf6938cb27511, y: 0x1815ac4c718e25f9f63c62ae84
4982317abbe69b7558a85180565bd8aac719917400672058904983699a8073d18e0745, infinity: Choice(0) })), spend_key: KeyPair(#3ef5cbf3d686a1fc) }
```

**Comment by justinmoon** [comment]
Does that load test tool spawn concurrent instances of `fedimint-cli`?

**Comment by douglaz** [comment]
> Does that load test tool spawn concurrent instances of `fedimint-cli`?

No, it just get some funds from the cli on the beginning then it spawn some Clients but each client is just doing one thing at once, there is no parallelism.

**Comment by dpc** [comment]
AFAIK this was caused by snapshot isolation breaking bugs in our rocksdb impl https://github.com/fedimint/fedimint/pull/5279. This has been fixed and is part of 0.4 and backported to 0.3.
